### PR TITLE
Add HTTPS template

### PR DESCRIPTION
**Pull request by tschuy**

### CODE REVIEW

File: templates/default/gwm_apache_vhost_https.conf.erb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<VirtualHost *:<%= @params[:server_port] || node['apache']['listen_ports'].first %>>
```
Comment (jordane) [inline templates/default/gwm_apache_vhost_https.conf.erb]: This won't do https..

Comment (tschuy) [inline templates/default/gwm_apache_vhost_https.conf.erb]: if we set server_port to 443 wouldn't it work?

Comment (jordane) [inline templates/default/gwm_apache_vhost_https.conf.erb]: Then why bother with the `||` at all?

File: recipes/apache.rb [hunk 2]
```diff
@@ -1,40 +1,48 @@
...
+
 web_app gwm['application_name'] do
-  template 'gwm_apache_vhost.conf.erb'
+  template template_name
   server_name node['hostname']
   server_aliases gwm['apache']['server_aliases']
-  cookbook "ganeti_webmgr"
+  cookbook 'ganeti_webmgr'
   server_name gwm['apache']['server_name']
+  server_port gwm['apache']['server_port']
```
Comment (tschuy) [inline recipes/apache.rb]: As it turns out this wasn't being passed before...

Comment (jordane) [inline recipes/apache.rb]: It's currently defaulting to 80, and then the template turns on SSL..

Comment (tschuy) [inline recipes/apache.rb]: The original template looks for @params[:server_port], but server_port wasn't being passed to the template. Now you'll be able to change it from port 80 :)

Comment (jordane) [inline recipes/apache.rb]: Right. My point is that this is a terrible default choice.

Comment (tschuy) [inline recipes/apache.rb]: so what should I put in the template? keep it as <VirtualHost *:<%= @params[:server_port] || node['apache']['listen_ports'].first %>> ? listen_ports won't have port 443 in it unless we run the apache::mod_ssl cookbook, and as of yet it's unclear to me -- do we have to run the mod_ssl cookbook? I would assume yes, but would like to confirm.

Comment (tschuy) [inline recipes/apache.rb]: or are you saying I should have something like this:

```

if gwm['https_enabled']
  template_name = 'gwm_apache_vhost_https.conf.erb'
  gwm['apache']['server_port'] = 443 unless gwm['apache']['server_port']
else
  template_name = 'gwm_apache_vhost.conf.erb'
end
```
?

Comment (jordane) [inline recipes/apache.rb]: I'm saying you should set `default['ganeti_webmgr']['apache']['server_port'] = 443`

Comment (tschuy) [inline recipes/apache.rb]: Shouldn't the default be set to 80, and then the cookbook user set it to 443? Or are we assuming that this cookbook will never be used to deploy HTTP service?

Comment (jordane) [inline recipes/apache.rb]: If the default is 80, then you **definitely** shouldn't set up HTTPS by default on port 80.

Comment (jordane) [inline recipes/apache.rb]: You should either:

1) Remove HTTPS by default and make it a flag
2) Make port 443 the default

Comment (tschuy) [inline recipes/apache.rb]: HTTPS isn't by default. In order to use HTTPS, the person using the cookbook:

1. Sets the node['ganeti_webmgr']['https_enabled'] flag to ``true``, from its default ``false``
2. Sets node['ganeti_webmgr']['apache']['server_port'] to ``443`` from its default ``80``

Comment (tschuy) [inline recipes/apache.rb]: I could set node['ganeti_webmgr']['apache']['server_port'] to nil by default, and have the cookbook decide the default based on whether https_enabled is true (as seen above with ``gwm['apache']['server_port'] = 443 unless gwm['apache']['server_port']``).

Comment (jordane) [inline recipes/apache.rb]: Oh, I missed that you have two templates.

Comment (jordane) [inline recipes/apache.rb]: It **would** make sense though, to set `gwm['apache']['server_port'] = 443` if `node['ganeti_webmgr']['https_enabled']` is true.

Comment (tschuy) [inline recipes/apache.rb]: Is what I'm doing safe? (currently line 35):

```
+if gwm['https_enabled']
+ template_name = 'gwm_apache_vhost_https.conf.erb'
+ gwm['apache']['server_port'] = 443 unless gwm['apache']['server_port']
+else
+ template_name = 'gwm_apache_vhost.conf.erb'
+end
+
```

Comment (jordane) [inline recipes/apache.rb]: Not really. Instead you should do
```ruby
if gwm['https_enabled']
  template_name = ...
  server_port = gwm['apache']['server_port'] == 80 ? 443 : gwm['apache']['server_port']
else
  server_port = gwm['apache']['server_port']
end
```

and then use the `server_port` var.

**Comment by armiller** [comment]
@tschuy this will need to get deployed. I tested with the lb in front and VNC doesn't work because its trying to open a web socket and throws a mixed content warning. As far as I can tell it randomizes web socket port so we can't proxy on specific port I.E. `8000`.

File: templates/default/gwm_apache_vhost_https.conf.erb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<VirtualHost *:<%= @params[:server_port] %>>
+
+    ServerName <%= @params[:server_name] %>
+    ServerAlias <% @params[:server_aliases].each do |a| %><%= a %> <% end %>
+    SSLEngine On
+
+
+    SSLCertificateFile "<%= @params[:app][:certificate_file] %>"
```
Comment (jordane) [inline templates/default/gwm_apache_vhost_https.conf.erb]: This isn't set anywhere?

Comment (tschuy) [inline templates/default/gwm_apache_vhost_https.conf.erb]: Ah, I suppose it should have a default.

Comment (jordane) [inline templates/default/gwm_apache_vhost_https.conf.erb]: The problem isn't so much  that it doesn't have a default, but it isn't passed to `@params` at all!

Comment (tschuy) [inline templates/default/gwm_apache_vhost_https.conf.erb]: @params[:app] is actually the gwm hash, which is node['ganeti_webmgr']. so @params[:app][:certificate_file] is node['ganeti_webmgr']['app']['certificate_file'].

Comment (jordane) [inline templates/default/gwm_apache_vhost_https.conf.erb]: That should be documented somewhere obvious then :)

Comment (tschuy) [inline templates/default/gwm_apache_vhost_https.conf.erb]: And both certificate_file and certificate_key_file are set by default to nil.

Comment (tschuy) [inline templates/default/gwm_apache_vhost_https.conf.erb]: Really what I want to do is rewrite this entire thing with python_webapp. 

Where would the optimal place to document this be? the apache config files are already using ``@params[:app]`` all over the place; I assume you want me to stick it in there?

Comment (jordane) [inline templates/default/gwm_apache_vhost_https.conf.erb]: Document it by setting the attributes in `attributes/<somewhere>.rb`.

Comment (jordane) [inline templates/default/gwm_apache_vhost_https.conf.erb]: Literally set the attributes to `nil` in the attributes files. Or some better default.

File: attributes/default.rb [hunk 4]
```diff
@@ -49,12 +49,17 @@
...
+default['ganeti_webmgr']['apache']['server_port'] = 80
 default['ganeti_webmgr']['apache']['server_name'] = node['hostname']
 default['ganeti_webmgr']['apache']['server_aliases'] = [node['fqdn']]
+default['ganeti_webmgr']['https_enabled'] = false
 
 default['ganeti_webmgr']['apache']['processes'] = 4
 default['ganeti_webmgr']['apache']['threads'] = 1
 
+default['ganeti_webmgr']['certificate_file'] = nil
+default['ganeti_webmgr']['certificate_key_file'] = nil
```
Comment (tschuy) [inline attributes/default.rb]: @jordane like this?

Comment (jordane) [inline attributes/default.rb]: +1

**Comment by jordane** [comment]
+1

**Pull request merged**